### PR TITLE
[symfony] Async commands

### PR DESCRIPTION
**Pull request by makasim**
# Async commands

The [enqueue/async-command](https://github.com/php-enqueue/async-command) package adds ability to run Symfony command async with ability to get output.

## Installation

```bash
$ composer require enqueue/async-command:^0.9.x-dev
```

## Configuration

```yaml
# config/packages/enqueue_async_commands.yaml

enqueue:
   async_commands: true
```

## Usage

```php
<?php

use Enqueue\Client\ProducerInterface;
use Enqueue\AsyncCommand\Commands;
use Enqueue\AsyncCommand\RunCommand;
use Symfony\Component\DependencyInjection\ContainerInterface;

/** @var $container ContainerInterface */

/** @var ProducerInterface $producer */
$producer = $container->get(ProducerInterface::class);

$producer->sendCommand(Commands::RUN_COMMAND, new RunCommand('debug:container'));
```

optionally you can get a command execution result:

```php
<?php

use Enqueue\Client\ProducerInterface;
use Enqueue\AsyncCommand\RunCommandResult;
use Enqueue\AsyncCommand\Commands;
use Enqueue\AsyncCommand\RunCommand;
use Symfony\Component\DependencyInjection\ContainerInterface;

/** @var $container ContainerInterface */

/** @var ProducerInterface $producer */
$producer = $container->get(ProducerInterface::class);

$promise = $producer->sendCommand(Commands::RUN_COMMAND, new RunCommand('debug:container'), true);

// do other stuff.

if ($replyMessage = $promise->receive(5000)) { 
    $result = RunCommandResult::jsonUnserialize($replyMessage->getBody());
    
    echo $result->getOutput();
}
```

[back to index](../index.md)

### CODE REVIEW

**Comment by wesleylancel** [comment]
@makasim Hi, any word on this being merged? We would love to use this :)

**Comment by makasim** [comment]
@wesleylancel It needs some testing, not sure I can find time to finish this in near future.

**Comment by temp** [comment]
Has enqueue/async-command been removed? The link in the first post produces a 404...
Would love to test this!

**Comment by stephanvierkant** [comment]
Same here!

**Comment by makasim** [comment]
I've created a repo `enqueue/async-command` and published it to packagist https://packagist.org/packages/enqueue/async-command

you can require it in your composer json like this

```json
{
  "require": {
    "enqueue/async-command": "dev-pr-250"
  }
}
```

**Comment by temp** [comment]
Thanks! Is there something like an example for a minimal setup/configuration for this? I'm new to enqueue, and just want to check wether or not to replace JmsJobQueue by this.

**Comment by makasim** [comment]
@temp the doc is in the PR https://github.com/php-enqueue/enqueue-dev/blob/ff43a41cee09ad81a86f14f488225de4031f55d9/docs/bundle/async_commands.md

**Comment by makasim** [comment]
and in PR's description.

**Comment by temp** [comment]
@makasim Hmm there is no *Bundle class, am I missing something?

File: pkg/enqueue-bundle/DependencyInjection/EnqueueExtension.php [hunk 1]
```diff
@@ -150,6 +151,15 @@ public function load(array $configs, ContainerBuilder $container)
             ]], $container);
         }
 
+        if ($config['async_commands']['enabled']) {
+            if (false == class_exists(AsyncCommandExtension::class)) {
```
Comment (temp) [inline pkg/enqueue-bundle/DependencyInjection/EnqueueExtension.php]: Won't the class constant lookup fail when the class is not available?

**Comment by temp** [comment]
Ok, got it. Had to manually edit the Configuration and the EnqueueExtension. Runs fine now, will report.

**Comment by rbrecht** [comment]
Maybe a small improvement we did:
We extended Enqueue\Client\Message, to have access to all message functions

```
<?php

namespace App\Async;

use Enqueue\Client\Message;

/**
 * Class RunCommand.
 */
final class RunCommand extends Message
{

    public function __construct(string $commandName, array $commandArguments = [], array $commandOptions = [], array $messageProperties = [], array $messageHeaders = [])
    {
        $argumentString = $this->createString($commandArguments);
        $optionsString = $this->createString($commandOptions);
        
        $message = sprintf('%s %s %s', $commandName, $argumentString, $optionsString);

        parent::__construct($message, $messageProperties, $messageHeaders);
    }

    private function createString(array $values): string
    {
        return implode(' ', $values);
    }
}
```

```
$command = new RunCommand('app:import-products', [$categoryId], ['--force']);
$command->setPriority(MessagePriority::HIGH);
$command->setDelay(60);
$this->producer->sendCommand(Commands::CONSOLE_COMMAND, $command);
```

**Comment by makasim** [comment]
@hslavich today (:

**Pull request merged**